src/connection_manager.rs

```diff
@@ -57,47 +57,61 @@ pub struct ConnectionManager {}
 
 impl ConnectionManager {
     /// Bootstrap to the network maintaining connections to several nodes.
-    pub async fn bootstrap(session: Session) -> Result<Session, Error> {
-        // &'static
+    pub async fn bootstrap(mut session: Session) -> Result<Session, Error> {
         trace!(
             "Trying to bootstrap to the network with public_key: {:?}",
             session.client_public_key()
         );
 
+        let (
+            endpoint,
+            _incoming_connections,
+            incoming_messages,
+            _disconnections,
+            bootstrapped_peer,
+        ) = session.qp2p.bootstrap().await?;
+
+        session.endpoint = Some(endpoint.clone());
+
         // Bootstrap and send a handshake request to
-        let (session, mut incoming_messages) = Self::get_section(session).await?;
-        // Receive the list of Elders we can then connect to
-        let session = Self::get_elders(session, &mut incoming_messages).await?;
+        let session = Self::get_section(session, Some(bootstrapped_peer)).await?;
+        let session = session
+            .listen_to_incoming_messages(incoming_messages)
+            .await?;
+
         // Let's now connect to all Elders
         let session = Self::connect_to_elders(session).await?;
 
-        session.listen_to_incoming_messages(incoming_messages).await
-    }
+        let mut we_have_keyset = false;
 
-    async fn get_elders(
-        session: Session,
-        incoming_messages: &mut IncomingMessages,
-    ) -> Result<Session, Error> {
-        if let Some((_src, message)) = incoming_messages.next().await {
-            match SectionInfoMsg::from(message) {
-                Ok(msg) => ConnectionManager::handle_sectioninfo_msg(msg, session).await,
-                Err(e) => Err(e.into()),
-            }
-        } else {
-            trace!("incoming_messages.next().await was None");
-            Err(Error::NotBootstrapped)
+        // bootstrap is not complete until we have pk set...
+        while !we_have_keyset {
+            use tokio::time::{delay_for, Duration};
+            delay_for(Duration::from_millis(500)).await;
+            we_have_keyset = session.section_key_set.lock().await.is_some();
         }
+
+        Ok(session)
     }
 
     /// Send a `Message` to the network without awaiting for a response.
     pub async fn send_cmd(
         msg: &Message,
-        endpoint: Endpoint,
-        elders: Vec<SocketAddr>,
+        session: &Session,
+        // endpoint: Endpoint,
+        // elders: Vec<SocketAddr>,
     ) -> Result<(), Error> {
         let msg_id = msg.id();
+        let endpoint = session.endpoint()?.clone();
+
+        let elders: Vec<SocketAddr>;
+
+        {
+            elders = session.elders.lock().await.iter().cloned().collect();
+        }
```

Comment:
```suggestion
        let elders = {
             session.elders.lock().await.iter().cloned().collect();
        }
```
Perhaps even without the braces. It should remain the same...